core/arch/arm/mm/tee_mmu.c

```diff
@@ -262,33 +359,74 @@ static void asid_free(unsigned int asid)
 	cpu_spin_unlock_xrestore(&g_asid_spinlock, exceptions);
 }
 
-TEE_Result tee_mmu_init(struct user_ta_ctx *utc)
+static TEE_Result map_kinit(struct user_ta_ctx *utc __maybe_unused)
 {
+#ifdef CFG_CORE_UNMAP_CORE_AT_EL0
+	struct mobj *mobj;
+	size_t offs;
+	vaddr_t va;
+	size_t sz;
+
+	thread_get_user_kcode(&mobj, &offs, &va, &sz);
+	return vm_map(utc, &va, sz, TEE_MATTR_PRX | TEE_MATTR_PERMANENT,
+		      mobj, offs);
+#else
+	return TEE_SUCCESS;
+#endif /*CFG_CORE_UNMAP_CORE_AT_EL0*/
+}
+
+TEE_Result vm_info_init(struct user_ta_ctx *utc)
+{
+	TEE_Result res;
 	uint32_t asid = asid_alloc();
 
 	if (!asid) {
 		DMSG("Failed to allocate ASID");
 		return TEE_ERROR_GENERIC;
 	}
 
-	utc->mmu = calloc(1, sizeof(struct tee_mmu_info));
-	if (!utc->mmu) {
+	utc->vm_info = calloc(1, sizeof(*utc->vm_info));
+	if (!utc->vm_info) {
 		asid_free(asid);
 		return TEE_ERROR_OUT_OF_MEMORY;
 	}
-	utc->mmu->asid = asid;
-	core_mmu_get_user_va_range(&utc->mmu->ta_private_vmem_start, NULL);
-	return TEE_SUCCESS;
+	TAILQ_INIT(&utc->vm_info->regions);
+	utc->vm_info->asid = asid;
+
+	res = map_kinit(utc);
+	if (res)
+		vm_info_final(utc);
+	return res;
 }
 
-static TEE_Result alloc_pgt(struct user_ta_ctx *utc __maybe_unused,
-			    vaddr_t base, vaddr_t end)
+void vm_info_get_user_range(struct user_ta_ctx *utc, vaddr_t *vstart,
+			    vaddr_t *vend)
+{
+	struct vm_region *r;
+
+	TAILQ_FOREACH(r, &utc->vm_info->regions, link)
+		if (!(r->attr & TEE_MATTR_PERMANENT))
+			break;
+
+	if (!r || TAILQ_EMPTY(&utc->vm_info->regions)) {
```

Comment:
Shouldn't `if (!r)` be sufficient here ?

Reply:
I think you're right, I'll update